src/transform-sdk/js/main.cc

```diff
@@ -0,0 +1,362 @@
+        if (errc) [[unlikely]] {
+            return std::unexpected(qjs::exception::make(
+              ctx, std::format("error writing record: {}", errc.message())));
+        }
+        return qjs::value::undefined(ctx);
+    }
+
+    std::expected<std::optional<redpanda::bytes_view>, qjs::exception>
+    extract_data(JSContext* ctx, const qjs::value& val) {
+        if (val.is_string()) {
+            _strings.emplace_back(val.string_data());
+            const auto& data = _strings.back();
+            return redpanda::bytes_view(data.view());
+        }
+        if (val.is_uint8_array()) {
+            auto data = val.uint8_array_data();
+            if (data.data() == nullptr) [[unlikely]] {
+                return std::unexpected(qjs::exception::current(ctx));
+            }
+            return redpanda::bytes_view(data.data(), data.size());
+        }
+        if (val.is_array_buffer()) {
+            auto data = val.array_buffer_data();
+            if (data.data() == nullptr) [[unlikely]] {
+                return std::unexpected(qjs::exception::current(ctx));
+            }
+            return redpanda::bytes_view(data.data(), data.size());
+        }
+        if (val.is_null() || val.is_undefined()) {
+            return std::nullopt;
+        }
+        record_data* record_data = _record_data->get_opaque(val);
+        if (record_data == nullptr) [[unlikely]] {
+            return std::unexpected(
+              qjs::exception::make(ctx, "unexpected type for record data"));
+        }
+        return record_data->data();
+    }
+
+private:
+    redpanda::record_writer* _writer;
+    qjs::class_factory<record_data>* _record_data;
+    std::vector<qjs::cstring> _strings; // a place to temporarily hold data.
+};
+
+qjs::class_factory<record_writer>
+make_record_writer_class(qjs::runtime* runtime) {
+    qjs::class_builder<record_writer> builder(
+      runtime->context(), "RecordWriter");
+    builder.method<&record_writer::write>("write");
+    return builder.build();
+}
+
+qjs::class_factory<record_data> make_record_data_class(qjs::runtime* runtime) {
+    qjs::class_builder<record_data> builder(runtime->context(), "RecordData");
+    builder.method<&record_data::json>("json");
+    builder.method<&record_data::text>("text");
+    builder.method<&record_data::array>("array");
+    return builder.build();
+}
+
+std::expected<qjs::value, qjs::exception> make_write_event(
+  JSContext* ctx,
+  const write_event& evt,
+  qjs::class_factory<record_data>* data_factory) {
+    qjs::value record = qjs::value::object(ctx);
+    std::expected<std::monostate, qjs::exception> result;
+    if (evt.record.key) {
+        result = record.set_property(
+          "key",
+          data_factory->create(std::make_unique<record_data>(*evt.record.key)));
+    } else {
+        result = record.set_property("key", qjs::value::null(ctx));
+    }
+    if (!result) [[unlikely]] {
+        return std::unexpected(result.error());
+    }
```

Comment:
I suspect there's a reason for this ceremony, but I don't know what it is

Reply:
It should never happen outside of alloc failures since we control the types here.